rust/blockstore/src/arrow/ordered_blockfile_writer.rs

```diff
@@ -136,6 +136,24 @@ impl ArrowOrderedBlockfileWriter {
 
         let mut split_block_deltas = Vec::new();
         for delta in inner.completed_block_deltas.drain(..) {
+            // Don't we split on-mutation (.set() calls)?
+            // Yes, but that is only a performance optimization. For correctness, we must also split on commit. Why?
+            //
+            // We need to defer copying old forked data until:
+            // - we receive a set()/delete() for a later key
+            // - we are committing the delta (it will receive no further writes)
+            //
+            // Because of this constraint, we cannot always effectively split on-mutation if the writer is over a forked blockfile. Imagine this scenario:
+            // 1. There is 1 existing block whose size == limit.
+            // 2. We receive a .set() for a key near the beginning of the existing block's key range
+            // 3. We copy N items from the old block into our new delta, then add the new KV pair.
+            // 4. At this point, the total size of the delta (materialized + pending forked data) is above the limit.
+            // 5. We would like to split our delta into two immediately after the newly-added key. However, this means that the right half of the split is empty (there is no materialized data), which violates a fundamental assumption made by our blockstore code. And we cannot materialize only the first key in the right half from the pending forked data because that would violate the above constraint. Additionally, this would result in a smaller-than-ideal block.
```

Comment:
In the ordered deltas, we actually enforce this by allowing single-key blocks. What is the "above constraint" and what is the perceived problem with "smaller-than-ideal" blocks? Can't you still end up in this case if no additional key is written after this set of steps?

Reply:
> In the ordered deltas, we actually enforce this by allowing single-key blocks.

sorry, don't understand what you're referring to

> What is the "above constraint"

We need to defer copying old forked data until:
- we receive a set()/delete() for a later key
- we are committing the delta (it will receive no further writes)

> what is the perceived problem with "smaller-than-ideal" blocks

I think is is irrelevant, I can update the comment.

> Can't you still end up in this case if no additional key is written after this set of steps?

sorry, not following, could you give an example?

Reply:
discussed in person